### ISSUE TITLE
Should `Port.validate` fail if `required=False` and receives `None` which does not match `valid_type`?

### ISSUE THREAD

**Issue by sphuber** [aiidateam/plumpy#226]
Original issue reported in [`aiida-common-workflows`](https://github.com/aiidateam/aiida-common-workflows/issues/244):

Consider the following example of a simple port that is not required and expects a string:
```python
In [1]: from plumpy import InputPort

In [3]: port = InputPort('test', valid_type=str, required=False)

In [4]: port.validate('string')

In [5]: port.validate(None)
Out[5]: plumpy.ports.PortValidationError("Error occurred validating port 'test': value 'test' is not of the right type. Got '<class 'NoneType'>', expected '<class 'str'>'")
```
When we validate `None` as a value, we hit the validation error, just as in the EOS workchain. Except that `None` should be a perfectly legal value if the port is optional. It seems weird to raise for this. The only reason I can think off to keep the current behavior is that it might be necessary to have an optional port but where an explicit `None` value is incorrect and so if `None` is passed explicitly (instead of the input being omitted entirely) you want the port to raise. But I am wondering if this is not an edge-case and in that case the develop should simply explicitly check for `None` in a validator or the process body and ignore it.

**Comment by sphuber** [comment]
@muhrin @chrisjsewell thoughts?

**Comment by muhrin** [comment]
Thanks Seb, I was actually just writing a response on the issue.

Essentially I agree with your analysis.  All of this comes down to an original choice I made which is best demonstrated with dictionaries (which input port hierarchies emulate).

Essentially, should this:
```
input = {
    'value': None
}
```
be considered the same as:
```
input = {}
```
i.e. should the lack of a key in a dictionary be equivalent to the key being there with a value of `None`.  This is equivalent to saying 'Should `None` be treated as a value or as the absence of a value?'

My instinct is that `None` should be treated as a value specifically to allow one to distinguish between the two dictionaries above.  Other languages deal with this using a third type, like `undef` which accounts for this case.  I think in plumpy we, internally at least, use an `UNSET` singleton for this case.

I don't remember if it's possible in plumpy to do (following from your example):
```
port.validate()
```
which should be `True` in this case.

To summarise, as it stands, I still prefer the origianl behaviour but I'm not opposed to the change.  I would just caution that this assumption has certain consensequence that aren't necessarily obvious and may appear in other parts of the code and so care should be taken to make sure the behvaiour is consistent throughout.

**Comment by chrisjsewell** [comment]
> Other languages deal with this using a third type, like undef

yeh never found a good/defacto solution to this in python

> Maybe we should change the signature for Port.validate to `value=UNSPECIFIED`

that makes sense to me
